### PR TITLE
T/113 b LivePosition, LiveRange and Selection

### PR DESCRIPTION
**Pull request by scofalik**
PR for #64 and #113 , introducing `LivePosition`, `LiveRange` and `Selection`

### CODE REVIEW

**Comment by pjasiun** [comment]
Selection attributes are missing.

**Comment by scofalik** [comment]
Since `Selection` would have to have same methods as `Node` to operate on attributes, I extracted "attributes interface" from `Node` class and created `AttributesInterfaceMixin` from it. This was the easiest way to do it without refactoring all `Node` usage and all `Node` tests. However, I am not sure, if introducing something like `AttributesSet` or `AttributesList` (like `NodeList`) would not be better. This would require rewriting code from things like `node.getAttr( x )` to `node.attrs.getAttr( x )`. Or rewriting methods like `getAttr( x ) { return this.attrs.getAttr( x ) }`. Then each class like `Node` or `Selection` would have to implement the whole set of those methods that would look the same.

If we stick with the solution I proposed, I have one more doubt. Right now it is difficult to use mixins that maintain their own properties, like `Set` of attributes in `AttributesInterfaceMixin`. We use lodash'es `.extend()` to extend "host" prototype: `utils.extend( Selection.prototype, AttributesInterfaceMixin )`. Unfortunately, when using this approach, we do shallow copy of properties, so:
```
const AttributesInterfaceMixin = {
  _attrs: new Set()
  // ...
}
```
`_attrs` is same for each instance of `Selection`. It's even worse when we use use getters as they are not copies as getter but as value:
```
let a = {
  _f: 0,
  get f() {
    return this._f * 10;
  }
  set f( _f ) {
    this._f = _f;
  }
}

let b = utils.extend( {}, a );

a.f = 10; // 100
b.f = 10;
b.f; // 10, not 100
```
We should do something about it if we want to continue using "mixins".

**Comment by pjasiun** [comment]
> Or rewriting methods like getAttr( x ) { return this.attrs.getAttr( x ) }

I prefer this way. This is how `NodeList` works now (not only elements, but also collection, can have a list of nodes). It is more flexible: you can publish only part of the attribute interface, change the behaviour, names of methods or have two sets of attributes. This may be a case for meta attributes where we will have `setMetaAttr`. Also KISS, `AttributesList` and `getAttr( x ) { return this.attrs.getAttr( x ) }` is simpler then `AttributesInterfaceMixin`, especially when you start hacking them.

BTW: `this._attrs` should be private if there is no reason to make it public.

**Comment by scofalik** [comment]
Pushed.

File: src/treemodel/selection.js [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+	/**
+	 * Unbinds all events bound by created {@link treeModel.LiveRange}s.
+	 *
+	 * @private
+	 * @method detachRanges
+	 * @memberOf {treeModel.Selection}
+	 */
+	function detachRanges() {
+		/*jshint validthis: true */
```
Comment (Reinmar) [inline src/treemodel/selection.js]:
```
/*jshint
 ><----- space missing here
```

File: src/treemodel/selection.js [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+	}
+
+	/**
+	 * Unbinds all events bound by created {@link treeModel.LiveRange}s.
+	 *
+	 * @private
+	 * @method detachRanges
+	 * @memberOf {treeModel.Selection}
+	 */
+	function detachRanges() {
```
Comment (Reinmar) [inline src/treemodel/selection.js]: Why is this function extracted? Can't we reuse `this.detach()` where this function is used?

File: src/treemodel/selection.js [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+			 */
+			this._ranges = [];
+
+			/**
+			 * Specifies whether the last added range was added as a backward or forward range.
+			 *
+			 * @private
+			 * @property {Boolean}
+			 */
+			this._lastRangeBackward = false;
```
Comment (Reinmar) [inline src/treemodel/selection.js]: Quoting @scofalik: in Firefox anchor and focus are set based on the last range. Since we don't want to cache these in a form of properties, we keep the last range's direction.

It makes sense that anchor is also taken from the last range when you make multi-range selection using `Cmd/Ctrl`. However, I've been wondering whether this will be also true in tables where you do a single drag to select a column. Turns out it is... I find it a bit odd, but it's legit.

I'm worried though that in case of some changes in the native selection (and such had been discussed at W3C meeting) this won't be true any more. If anchor and focus will be decoupled from the last range what will we need to change in the API and implementation to make it work? Also, what if beside anchor and focus another "point" will be added to the selection API (this actually isn't just a theory, because I remember Ryosuke Niwa mentioning such a possibility).

Comment (Reinmar) [inline src/treemodel/selection.js]: Side note – does it have to be "backward"? I prefer if we kept a positive properties (makes thinking easier :D).

File: src/treemodel/selection.js [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+		 * Unbinds all events previously bound by this selection, including events bound by created {@link treeModel.LiveRange}s.
+		 */
+		detach() {
+			detachRanges.call( this );
+		}
+
+		/**
+		 * @see {@link treeModel.AttributeList#getAttr}
+		 */
+		getAttr( key ) {
```
Comment (Reinmar) [inline src/treemodel/selection.js]: Why can't we make _attrs public?

File: src/treemodel/position.js [hunk 5]
```diff
@@ -355,6 +345,58 @@ CKEDITOR.define( [ 'treemodel/rootelement', 'utils', 'ckeditorerror' ], ( RootEl
 		}
 
 		/**
+		 * Checks whether this position is touching given position. Positions touch when there are no characters
+		 * or empty nodes in a range between them. Technically, those positions are not equal but in many cases
+		 * they are very similar or even indistinguishable when they touch.
```
Comment (Reinmar) [inline src/treemodel/position.js]: "when they touch"... https://www.youtube.com/watch?v=z13qnzUQwuI

File: src/treemodel/position.js [hunk 6]
```diff
@@ -406,7 +448,17 @@ CKEDITOR.define( [ 'treemodel/rootelement', 'utils', 'ckeditorerror' ], ( RootEl
...
+			return new this( parent.root, path );
+		}
+
+		/**
+		 * Creates and returns a new instance of Position, which is equal to passed position.
+		 *
+		 * @param {treeModel.Position} position Position to be cloned.
+		 * @returns {treeModel.Position}
+		 */
+		static createFromPosition( position ) {
```
Comment (Reinmar) [inline src/treemodel/position.js]: `Position.createFromPosition()`? :|

File: src/treemodel/position.js (hunk 5 above)
Comment (scofalik) [inline src/treemodel/position.js]: At least position can touch this. https://www.youtube.com/watch?v=otCpCn0l4Wo

File: src/treemodel/selection.js (hunk 2 above)
Comment (scofalik) [inline src/treemodel/selection.js]: It is extracted in case if `Selection` bind it's own events. Detaching ranges is used in multiple places, where we would not like to detach `Selection` own events.

But right now `Selection` does not bind any events and I am not sure it ever will so it might be too hasty to extract this part of code.

File: src/treemodel/selection.js (hunk 3 above)
Comment (scofalik) [inline src/treemodel/selection.js]: If `anchor` and `focus` setting method changes we would have to deprecate `isBackward` and `isLastBackward` parameters. Then we would have to implement new way of setting them - they might be set automatically depending on new logic or set in explicit way. We might need to change getters to normal properties. These aren't big changes really. There are three things that we need to remember:
1) Right now our implementation is correct compared to Firefox. I have to admit that I haven't read W3C docs on how it should work, though. I trust that Firefox did it right. However, reading their docs, they don't have any explicit way to set the selection direction. You do it by setting range/collapsing selection and then by extending it. So they do it more behind the scenes.
2) It is rare that the user cares which way selection is set.
3) If anything becomes to complicated to change we still have the flexibility because this `Selection` class "lives" only in Tree Model space. It should have useful API so users can easily operate on it, but it will be "translated" into DOM Selection so we might do some hacking there. This is of course only last resort, if ever things get too complicated. Of course the primary goal is to be standards compilant.
BTW. If you would like me to make our `Selection` more like DOM one (API) let me know and I will change/add some methods.

As far as "backward" is concerned - I don't know, I think I just like when optional parameters default to `false`. For me it comes natural when I write a function. I can change it if you want :P.

File: src/treemodel/selection.js (hunk 4 above)
Comment (scofalik) [inline src/treemodel/selection.js]: You can read comments above on how the code ended up this way. I had to decide whether to:

- keep `Node` API same so this way we do not change code using `Node` and tests, and it (maybe) feels more natural that we "set" attribute on a node, or check whether a node has attribute

OR

- make `_attrs` of `Node` public and use it's API (see `AttributeList`), Then we would have to change whole API but we won't have silly functions like `getAttr( key ) { return this._attrs.getAttr( key ); }`.

No matter which way we go I think that everywhere we use attributes, we should use same approach so the user don't get confused. Personally, I think I prefer to use it like: `node.attrs.has( key );`, `selection.attrs.add( attr )` etc. It spares us a lot of code when it comes to those "interface" functions and also when it comes to test because we don't have to repeat almost exactly same tests for both `AttributeList` and every class that uses it.

Honestly, the more I think of it, the more I am decided to refactor code that uses `Node` and change this attributes API.

File: src/treemodel/position.js (hunk 6 above)
Comment (scofalik) [inline src/treemodel/position.js]: This was explained during F2F discussion. More on this subject is here: https://github.com/ckeditor/ckeditor5-core/issues/64#issuecomment-163213464 (though I have to admit it may be difficult to understand). Long story short: `.this.clone()` is not going to work and we either introduce a parameter to `clone` method or use static creators. I think this solution is much cleaner and explicit regarding what happens behind scenes.

BTW. I was for `Position.clone` and `Range.clone` names so if you prefer those over `createFromPosition` and `createFromRange` we can discuss it. @pjasiun was against them.

**Comment by scofalik** [comment]
I pushed commit with missing space :D

File: src/treemodel/position.js (hunk 6 above)
Comment (Reinmar) [inline src/treemodel/position.js]: Taken the confusing behaviour which clone() would have, I think it's reasonable to stay with `Position.createFromPosition()` and `LivePosition.createFromPosition()`.

Comment (scofalik) [inline src/treemodel/position.js]: To clear it up: I've meant static method named `.clone`, not instance method.

File: src/treemodel/selection.js (hunk 2 above)
Comment (Reinmar) [inline src/treemodel/selection.js]: Hm... since I don't understand this explanation it may be a good idea to move this code to `Selection.detach()` :D. Also, not keeping the logic totally private allows you to change it if you choose to derive from this class. E.g. `setRanges()` uses `detachRanges()` now. If you'd create `Selection`'s subclass and change how the `detach()` works you would not affect how `setRanges()` works even though it seems to require the same logic.

It's theory and most likely won't happen in this case, but it's a general argument for keeping stuff in the prototype.

File: src/treemodel/position.js (hunk 6 above)
Comment (Reinmar) [inline src/treemodel/position.js]: Ah... Then I don't have a strong opinion :) Keeping `createFromXXX` will be consistent with other methods which we already have and changing to `clone` will make it look better. Let's keep what we have now.

File: src/emittermixin.js [hunk 7]
```diff
@@ -15,6 +15,13 @@
 CKEDITOR.define( [ 'eventinfo', 'utils' ], ( EventInfo, utils ) => {
 	const EmitterMixin = {
 		/**
+		 * Saves how many callbacks has been already added. Does not decrement when callback is removed.
+		 * Used internally as a unique id for a callback.
+		 * @private
+		 */
+		_counter: 0,
```
Comment (Reinmar) [inline src/emittermixin.js]: There's too high chance of a conflict with classes private property. Can't we keep this as a private *variable* in this scope? If not, then maybe `_emitterMixinCallbackCounter` :D.

File: src/treemodel/document.js [hunk 8]
```diff
@@ -76,6 +85,16 @@ CKEDITOR.define( [
...
+		 * Returns current selection done on this document.
+		 *
+		 * @readonly
+		 * @property {treeModel.Selection}
+		 */
+		get selection() {
+			return this._selection;
+		}
+
+		/**
```
Comment (pjasiun) [inline src/treemodel/document.js]: Why not simply `this.selection = new Selection();`? It is marked as `@readonly` anyway.

File: src/treemodel/liveposition.js [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+
+			/**
+			 * Flag representing LivePosition stickiness. LivePosition might be sticking to previous node or next node.
+			 * Whenever some nodes are inserted at the same position as LivePosition, `stickiness` is checked to decide if
+			 * LivePosition should be moved. Similar applies when a range of nodes is moved and one of it's boundary
+			 * position is same as LivePosition. Accepted values are {@link #STICKS_TO_PREVIOUS} and {@link #STICKS_TO_NEXT}.
+			 *
+			 * @type {Number}
+			 */
+			this.stickiness = stickiness ? stickiness : STICKS_TO_NEXT;
```
Comment (Reinmar) [inline src/treemodel/liveposition.js]: `stickiness = stickiness || STICKS_TO_NEXT` will be simpler.

File: src/treemodel/liveposition.js [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+		 */
+	}
+
+	/**
+	 * Binds this LivePosition to the {@link treeModel.Document} that owns this position {@link treeModel.RootElement root}.
+	 *
+	 * @private
+	 * @method bindWithDocument
+	 */
+	function bindWithDocument() {
```
Comment (Reinmar) [inline src/treemodel/liveposition.js]: Again, not sure why is this kept outside of a proto. Even if this will need to be reused by some other code in the future this can still be a private method. You will not need that JSHint comments and using `.call()`.

File: src/treemodel/liveposition.js [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+		 * @constructor
+		 */
+		constructor( root, path, stickiness ) {
+			super( root, path );
+
+			/**
+			 * Flag representing LivePosition stickiness. LivePosition might be sticking to previous node or next node.
+			 * Whenever some nodes are inserted at the same position as LivePosition, `stickiness` is checked to decide if
+			 * LivePosition should be moved. Similar applies when a range of nodes is moved and one of it's boundary
+			 * position is same as LivePosition. Accepted values are {@link #STICKS_TO_PREVIOUS} and {@link #STICKS_TO_NEXT}.
```
Comment (pjasiun) [inline src/treemodel/liveposition.js]: Example would be useful. For the person who have no idea what the problem is this explanation may be not enough. And it is very simple to explain with the ASCII image.

File: src/treemodel/liverange.js [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+CKEDITOR.define( [
+	'treemodel/range',
+	'treemodel/liveposition',
+	'emittermixin',
+	'utils'
+], ( Range, LivePosition, EmitterMixin, utils ) => {
+	/**
+	 * LiveRange is a Range in the Tree Model that updates itself as the tree changes. It may be used as a bookmark.
+	 * **Note:** Be very careful when dealing with LiveRange. Each LiveRange instance bind events that might
+	 * have to be unbound. Use {@link #detach} whenever you don't need LiveRange anymore.
```
Comment (pjasiun) [inline src/treemodel/liverange.js]: What is the difference between `liveRange` and range with two `livePosition`s?

File: src/treemodel/liverange.js [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+			( event, type, changes ) => {
+				fixBoundaries.call( this, type, changes.range, changes.sourcePosition );
+			},
+			this
+		);
+	}
+
+	/**
+	 * LiveRange is partially updated by "automatic" updates of LivePositions that are boundaries of LiveRange.
+	 * However there are cases when the boundaries have to be fixed because they end up in wrong places.
```
Comment (pjasiun) [inline src/treemodel/liverange.js]: "there are cases" What cases? Example would be useful. You can put such example in the class description to show what is the difference between `liveRange` and range of `livePosition`s.

File: src/treemodel/attributelist.js [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+
+			this._attrs.add( attr );
+		}
+
+		/**
+		 * Removes all attributes and sets passed attributes.
+		 *
+		 * @param {Iterable.<treeModel.Attribute>} attrs Array of attributes to set.
+		 */
+		setAttrsTo( attrs ) {
```
Comment (pjasiun) [inline src/treemodel/attributelist.js]: Do we need this such method at all?

File: src/treemodel/range.js [hunk 15]
```diff
@@ -25,33 +25,41 @@ CKEDITOR.define( [ 'treemodel/position', 'treemodel/positioniterator', 'utils' ]
 			 *
 			 * @property {treeModel.Position}
 			 */
-			this.start = start;
+			this.start = Position.createFromPosition( start );
```
Comment (pjasiun) [inline src/treemodel/range.js]: It means that if you create a static range from the live position the live position is transformed into static position? This should be mentioned in the range class or constructor documentation.

File: src/treemodel/selection.js (hunk 3 above)
Comment (Reinmar) [inline src/treemodel/selection.js]: The biggest problem here is that there's no spec. The [current one](http://w3c.github.io/selection-api/) does specify only single range selections. There was also Aryeh Gregor's work but since it was based on how things work in the browsers it either described FF's or other browser's behaviour :D.

Anyway, the point is that in the future the native selection may have more features than the current one and if there's a an API design that could accommodate these features in the future, then it's the way to go. And I think that the API you proposed is good because so far it works automatically (anchor and focus are calculated on the fly). If we'll need to introduce some property or to allow setting some property manually (so it's not calculated based on the last range), then we can do that by extending the API.

File: src/treemodel/liveposition.js (hunk 10 above)
Comment (scofalik) [inline src/treemodel/liveposition.js]: During one of discussions we decided that it is not a bad idea to keep private methods really private. I proposed to do it like t his and I followed that pattern. Everywhere in code where is private and used only by the class I did it this way.

By definition, private function is not available not only outside class but also for child-classes. Child-class should not extend private function. Child-class may extend public or protected function and call `super.func()` then make additional logic for it own.

So, to sum it up, it all depends on how we define what is private function and how we decide to write them.

File: src/treemodel/liverange.js (hunk 12 above)
Comment (scofalik) [inline src/treemodel/liverange.js]: Recently I changed `Range` constructor so it creates it's own `Position` instances from the passed variables. So you can create a range using `LivePosition`s but positions of newly created range won't be live.

This approach has much more sense when it comes to creating `LiveRange` and `Selection`. They create their own "live" parts, manage them and detach.

File: src/treemodel/liverange.js (hunk 13 above)
Comment (scofalik) [inline src/treemodel/liverange.js]: As I've written above, there is no such thing like `Range` with two `LivePosition`s. Instead I could show a difference between `LiveRange` and two loose `LivePosition`s.

File: src/treemodel/attributelist.js (hunk 14 above)
Comment (scofalik) [inline src/treemodel/attributelist.js]: Well it depends how we want to handle some cases. Let's say that DOM Selection has been changed and we need to set new attributes for `Selection`. We could do one of two:

* create new 'AttributeList` and incrementaly add all attributes using `setAttr`, or
* create new `AttributeList` and pass all attirbutes in constructor and asign it to the selection, or
* use this method to assign all attributes at once.

It depends whether we want members that are `AttributeList` to be readonly or not, how we want to add multiple attributes and if we want `removeAllAttrs` method.

BTW. This method should be used in `AttributeList` constructor.

File: src/treemodel/range.js (hunk 15 above)
Comment (scofalik) [inline src/treemodel/range.js]: Isn't it enough that property is of type `treeModel.Position` ?

File: src/treemodel/position.js (hunk 6 above)
Comment (scofalik) [inline src/treemodel/position.js]: So I am leaving it.

File: src/treemodel/selection.js (hunk 3 above)
Comment (scofalik) [inline src/treemodel/selection.js]: Okay so since noone provided a real problem with this, I am leaving it as is.

File: src/treemodel/liveposition.js (hunk 10 above)
Comment (Reinmar) [inline src/treemodel/liveposition.js]: Hm... I can't recall this discussion. Maybe I wasn't there :D.

Anyway, you're actually right that if it's private, not protected, then it shouldn't be possible to modify it. So I'm OK now :)

Comment (scofalik) [inline src/treemodel/liveposition.js]: https://github.com/ckeditor/ckeditor5-design/issues/84

Here it is. Well, at the end we haven't agreed how it should be done. So I have been using my proposal :P

File: src/treemodel/selection.js (hunk 4 above)
Comment (Reinmar) [inline src/treemodel/selection.js]: So it seems that it requires refactoring. Those method are duplicated right now in `Node` and `Selection` only for one reason – because initially we had them directly in `Node` because attribute list was integral part of it. Now it's not, so all those methods are unnecessary.

Comment (scofalik) [inline src/treemodel/selection.js]: Will be done in separate ticket.

File: src/treemodel/range.js (hunk 15 above)
Comment (scofalik) [inline src/treemodel/range.js]: I added a mention in the constructor doc.

**Comment by scofalik** [comment]
I've pushed (hopefully) final commit with changes discussed in this PR.
